### PR TITLE
feat: Custom Terraform Provider section

### PR DESCRIPTION
**Pull request by baserrato**
DevOps Bootcamp Custom Terraform Provider Section 6.1.4

### CODE REVIEW

**Comment by baserrato** [comment]
Work in Progress, made draft pr to have an easier way to show the section and get feedback for what needs to be improved.

File: 6-infrastructure-configuration-management/6.1.4-terraform-providers.md [hunk 1]
```diff
@@ -14,49 +14,199 @@ Throughout this section it will be helpful to refer to the [Custom Framework Pro
...
+      # example configuration here
+    }
+    ```
 
 7. Run `terraform plan` to ensure Terraform can find and use your provider.
+
+### API Overview
+
+This section of the boot camp will focus on using an ephemeral API
+([API ReadMe](https://github.com/liatrio/devops-bootcamp/blob/a331c9598fdd5187eb324a9b54578e45b7a5dfb9/examples/ch6/devops-api/readMe.md))
```
Comment (Pactionly) [inline 6-infrastructure-configuration-management/6.1.4-terraform-providers.md]: Is this link correct? It probably shouldn't link to a specific commit, but if it does need to, the readme appears to be empty in this commit.

File: 6-infrastructure-configuration-management/6.1.4-terraform-providers.md [hunk 2]
```diff
@@ -14,49 +14,199 @@ Throughout this section it will be helpful to refer to the [Custom Framework Pro
...
+
+?> For this section ignore how the datasource is implimented in the tutorial as it is an incredibly uncommon approach, your datasource should be 1:1 to another resource of the same type.  Datasources are used to get information about resources not managed by terraform, and given that they arent being managed, you will lack certian information about them such as an ID so you will need to look at the API for another way to grab the desired resources. 
+
+?> Also think about what should happen when Terraform goes to fetch information about a resource it has inside its state but cant find in the API.  What does that mean and how should it be corrected?
+
+### Exercise 5: Implementing Dev, Ops, DevOps resource and datasource
+
+Implementing the next resoruces are going to be similar to the implementation of the engineer resoruce. Refer back to the section [API Overview](#api-overview)
+
+1. Begin with implementing the read functionality
```
Comment (Pactionly) [inline 6-infrastructure-configuration-management/6.1.4-terraform-providers.md:202]: Is there more steps intended to be added here?

File: 6-infrastructure-configuration-management/6.1.4-terraform-providers.md [hunk 3]
```diff
@@ -14,49 +14,199 @@ Throughout this section it will be helpful to refer to the [Custom Framework Pro
...
+   1. Think about what the Schema should look like:  
+    - What does the data the API returns going to look like?
+    - What variables does the API need vs what variables will be given to us?
+
+   1. Impliment all of the required CRUD operations using the tutorial as a reference.  
+    - In the documentation they use methods within their client wrapper to perform the API calls:
+        - This is time where you should plan how you want to organize these for the rest of your resources.
+        - Recommendation: Making a wrapper as done in the tutorial is recommended to make the code easier to read and understand.
+
+?> For this section ignore how the datasource is implimented in the tutorial as it is an incredibly uncommon approach, your datasource should be 1:1 to another resource of the same type.  Datasources are used to get information about resources not managed by terraform, and given that they arent being managed, you will lack certian information about them such as an ID so you will need to look at the API for another way to grab the desired resources.
```
Comment (Pactionly) [inline 6-infrastructure-configuration-management/6.1.4-terraform-providers.md:194]: implimented -> implemented

**Review by Pactionly — CHANGES_REQUESTED** [review]

File: 6-infrastructure-configuration-management/6.1.4-terraform-providers.md (hunk 3 above)
Comment (baserrato) [reply 6-infrastructure-configuration-management/6.1.4-terraform-providers.md]: fixed the spelling error with implement

File: 6-infrastructure-configuration-management/6.1.4-terraform-providers.md (hunk 1 above)
Comment (baserrato) [reply 6-infrastructure-configuration-management/6.1.4-terraform-providers.md]: Fixed empty document link with this [commit](https://github.com/liatrio/devops-bootcamp/pull/214/commits/5b1a5880be11f002a41753756d4fd3b3aa8414e5). Still links to a specific commit as a permalink for the API readme.

File: 6-infrastructure-configuration-management/6.1.4-terraform-providers.md [hunk 4]
```diff
@@ -14,49 +14,199 @@ Throughout this section it will be helpful to refer to the [Custom Framework Pro
...
+   load any plugins automatically
+   exit status 1
+   ```
 
 4. Build a binary for your provider `go build -o terraform-provider-devops-bootcamp`
 
 5. Edit `~/.terraformrc` and add the follow dev override to use the local version of your provider:
-  ```
-  provider_installation {
+    ```go
```
[the remark this answers is not in the document]
Comment (baserrato) [reply 6-infrastructure-configuration-management/6.1.4-terraform-providers.md]: Changed terraform code block to hcl. You're right it should be hcl not go.

File: 6-infrastructure-configuration-management/6.1.4-terraform-providers.md (hunk 2 above)
Comment (baserrato) [reply 6-infrastructure-configuration-management/6.1.4-terraform-providers.md:202]: Exercise 5 for the Dev, Ops and DevOps resource and data sources will have more steps. This is more of a repeat of engineer so we're wanting to ensure Exercise 4 (the engineer resource and data source) has more details explaining how to make the resource and test it.

**Review by Pactionly — APPROVED** [review]

**Pull request merged**